### ISSUE TITLE
mypy sometimes emits fewer errors when running on multiple files at once

### ISSUE THREAD

**Issue by marc-legendre** [python/mypy#14348]
**Bug Report**

Hello there,

I am witnessing (rare) cases where mypy emits fewer errors when running on multiple files at the same time than when running on each file individually.

More specifically, I noticed that mypy can fail to report an error when an import for a submodule is missing a one file but present in another file.

I failed to find anything relevant in the documentation or in the existing issues, so I thought I'd let you know. Below is my attempt at a minimal reproducible example.

**To Reproduce**

(Since multiple files are needed, here is an archive with the reproduction case: [repro.zip](https://github.com/python/mypy/files/10304465/repro.zip))

Given a module `mymodule` with nested modules `a` and `b`, and this file `foo.py`:

```py
import mymodule.a


def blah() -> None:
    mymodule.b.fn()


blah()
```

As expected, when called on `foo.py`,  `mypy` emits an error:

```
$ mypy --cache-dir /dev/null foo.py
foo.py:5: error: Module has no attribute "b"
Found 1 error in 1 file (checked 1 source file)
```

yet if I add to mypy's command line a second file which does import `mymodule.b`, say `bar.py`:

```py
import mymodule.b
```

then `mypy` does not emit the error any more, even though I would expect it to:

```
$ mypy --cache-dir /dev/null foo.py bar.py
Success: no issues found in 2 source files
```

**Your Environment**

I used mypy 0.991 on Ubuntu 22.04 with python 3.10.6; I can reproduce this behavior with `master`, as well as python 3.8, and older versions of mypy.
